pkg/replication/replication.go

```diff
@@ -1,195 +1,173 @@
+func (r Replicator) CheckPreparers() error {
+	for _, host := range r.Nodes {
+		_, _, err := r.Store.Pod(kp.RealityPath(host, "p2-preparer"))
+		if err != nil {
+			return util.Errorf("Host %q does not have a preparer", host)
 		}
 	}
-	return ret, err
-}
-
-type IntentStore interface {
-	SetPod(node string, manifest pods.Manifest) (time.Duration, error)
+	return nil
 }
 
-func NewReplicator(manifest pods.Manifest, allocated allocation.Allocation) *Replicator {
-	sha, _ := manifest.SHA()
-	replicator := Replicator{
-		MinimumNodes: 1,
-		Manifest:     manifest,
-		Allocation:   allocated,
-		Logger: logging.NewLogger(logrus.Fields{
-			"pod": manifest.ID(),
-			"sha": sha,
-		}),
-		NodePauseTime: 10 * time.Second,
+// Attempts to claim a lock on every host being deployed to.
+func (r Replicator) LockHosts(lock kp.Lock) error {
+	for _, host := range r.Nodes {
+		lockPath := kp.LockPath(host, r.Manifest.ID())
+		err := lock.Lock(lockPath)
+
+		if _, ok := err.(kp.AlreadyLockedError); ok {
+			holder, err := r.Store.LockHolder(lockPath)
+			if err != nil {
+				return util.Errorf("Lock already held for %q, could not determine holder due to error: %s", lockPath, err)
+			} else if holder == "" {
+				// we failed to acquire this lock, but there is no outstanding
+				// holder
+				// this indicates that the previous holder had a LockDelay,
+				// which prevents other parties from acquiring the lock for a
+				// limited time
+				return util.Errorf("Lock for %q is blocked due to delay by previous holder", lockPath)
+			} else {
+				return util.Errorf("Lock for %q already held by lock %q", lockPath, holder)
+			}
+		} else if err != nil {
+			return err
+		}
 	}
-	return &replicator
+	return nil
 }
 
-// Enact will try forever to upgrade the given nodes to the version specified and ensure that the minimum
-// number of healthy nodes is met or exceeded. If an error occurs while attempting to roll out, it will
-// log the error and retry. This loop can be terminated by closing the provided stop channel.
-func (repl *Replicator) Enact(store IntentStore, serviceChecker ServiceChecker, stop <-chan struct{}) {
-	if repl.MinimumNodes >= len(repl.Allocation.Nodes) {
-		repl.Logger.WithFields(logrus.Fields{
-			"minimum":   repl.MinimumNodes,
-			"allocated": len(repl.Allocation.Nodes),
-		}).Errorln("Minimum nuber of nodes cannot meet or exceed allocated")
-		return
-	}
-	repl.Logger.NoFields().Infoln("Enacting replication")
-	status, err := serviceChecker.LookupHealth(repl.Manifest.ID())
-	if err != nil {
-		repl.Logger.WithField("err", err).Errorln("Couldn't communicate with health checker on first run")
-		return
+// Execute the replication.
+// note: error management could use some improvement, errors coming out of
+// updateOne need to be scoped to the node that they came from
+func (r Replicator) Enact(errCh chan<- error, quitCh <-chan struct{}) {
+	nodeQueue := make(chan string)
+	done := make(chan string)
+	innerQuit := make(chan struct{})
```

Comment:
yeah the trick here is to terminate an unknown number of goroutines without having to actually send on the channel (which might block if you send more values than there are receivers).